Fix part of #8016: Migration of Exploration rights Service from AngularJS to Angular

heyimShivam opened the pull request:
## Overview
AngularJS to Angular Migration of service exploration rights

1. This PR fixes or fixes part of #8016.
2. This PR does the following: AngularJS to Angular Migration of service

## Essential Checklist

- [X] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [X] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [X] The PR is made from a branch that's **not** called "develop".

## Proof that changes are correct


https://user-images.githubusercontent.com/77006897/142734020-1cfa5ac9-de42-4bc7-aa09-71d1f3db52d2.mp4



![proof](https://user-images.githubusercontent.com/77006897/142463193-1142769e-ad49-444f-bd02-13d5bf964363.png)

## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#instructions-for-making-a-code-change).
- Oppiabot will notify you when you don't add a PR_CHANGELOG label. If you are unable to do so, please @-mention a code owner (who will be in the Reviewers list), or ask on [Gitter](https://gitter.im/oppia/oppia-chat).
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.
- Oppiabot can assign anyone for review/help if you leave a comment like the following: "{{Question/comment}} @{{reviewer_username}} PTAL"
- Some of the e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-your-build-fails).

heyimShivam committed:
first commit

heyimShivam committed:
80 max-len

heyimShivam committed:
year Change

heyimShivam committed:
semicolon error by checks solved

heyimShivam commented:
@ashutoshc8101 @EricZLou  @srijanreddy98 @aishwary023 @gp201 @seanlip PTAL

seanlip commented:
@heyimShivam We typically start with 1-2 core codeowners, then the rest. So, just assigning @ashutoshc8101 and @srijanreddy98 for now.

Also please could you make your title clearer about what the PR is doing? It should start with a verb (after the "fix ..." part).

heyimShivam commented:
okay

ashutoshc8101 commented on core/templates/pages/exploration-editor-page/services/exploration-rights-backend-api.service.ts:
```diff
@@ -0,0 +1,262 @@
+  providedIn: 'root'
+})
+export class ExplorationRightsService {
+  ownerNames: string[] = undefined;
+  editorNames: string[] = undefined;
+  voiceArtistNames: string[] = undefined;
+  viewerNames: string[] = undefined;
+  _status: string = undefined;
+  _clonedFrom: string = undefined;
+  _isCommunityOwned: boolean = undefined;
+  _viewableIfPrivate: boolean = undefined;
+
```
Assigned to `undefined` is not required. Can you please make all variable prefixed with `_` private.

heyimShivam replied:
Done

ashutoshc8101 commented on core/templates/pages/exploration-editor-page/services/exploration-rights-backend-api.service.ts:
```diff
@@ -0,0 +1,262 @@
+    private alertsService: AlertsService,
+    private httpClient: HttpClient,
+    private explorationDataService: ExplorationDataService,
+  ) { }
```
```suggestion
  ) {}
```

and we follow convention of naming httpClient variable as http.

heyimShivam replied:
Done

ashutoshc8101 commented on core/templates/pages/exploration-editor-page/services/exploration-rights-backend-api.service.ts:
```diff
@@ -0,0 +1,262 @@
+  }
+
+  isPublic(): boolean {
+    return this._status === constants.ACTIVITY_STATUS_PUBLIC;
```
Please use `AppConstants` instead of `constants`.

heyimShivam replied:
Done

ashutoshc8101 commented on core/templates/pages/exploration-editor-page/services/exploration-rights-backend-api.service.ts:
```diff
@@ -0,0 +1,262 @@
+  }
+
+  makeCommunityOwned(): Promise<void> {
+    let requestUrl = (
```
Please try to use const for URLs which are not changing.

heyimShivam replied:
Done

ashutoshc8101 commented on core/templates/pages/exploration-editor-page/services/exploration-rights-backend-api.service.ts:
```diff
@@ -0,0 +1,262 @@
+        data.rights.status, data.rights.cloned_from,
+        data.rights.community_owned, data.rights.viewable_if_private);
+    }).catch(() => {
+      this.init(
+        undefined, undefined, undefined, undefined,
+        undefined, undefined, undefined, undefined);
```
Try to use `null` instead of `undefined`.

heyimShivam replied:
Done

ashutoshc8101 reviewed (changes requested):
@heyimShivam Took a pass. PTAL

heyimShivam committed:
ashu changes

ashutoshc8101 commented on core/templates/pages/exploration-editor-page/services/exploration-rights-backend-api.service.spec.ts:
```diff
@@ -0,0 +1,445 @@
+    });
+    als = TestBed.inject(AlertsService);
+    csrfService = TestBed.inject(CsrfTokenService);
+    ers =
+      TestBed.inject(ExplorationRightsService);
```
Can you do the assignment in a single line?

heyimShivam replied:
Done

ashutoshc8101 commented on core/templates/pages/exploration-editor-page/services/exploration-rights-backend-api.service.spec.ts:
```diff
@@ -0,0 +1,445 @@
+    httpTestingController.verify();
+  });
+
+  it('should correctly initializes the service', function() {
```
Please use arrow functions here.

heyimShivam replied:
Done

ashutoshc8101 reviewed (changes requested):
Just few nits and please drop the `-backend-api` suffix in the file name.
This will also cause you to revert all the changes made to require/imports.

ashutoshc8101 commented:
@heyimShivam Please add a video showing that the functionality still works the same.

heyimShivam commented:
> Just few nits and please drop the `-backend-api` suffix in the file name. This will also cause you to revert all the changes made to require/imports.

I think while pushing the code it gives an error ` You are not allowed to create http requests from files that are not backend api services.eslint(oppia/disallow-httpclient)  `

heyimShivam committed:
done

heyimShivam committed:
solved e2e test case

ashutoshc8101 commented on core/templates/pages/exploration-editor-page/services/exploration-rights-backend-api.service.ts:
```diff
@@ -0,0 +1,262 @@
+@Injectable({
+  providedIn: 'root'
+})
+export class ExplorationRightsService {
+  ownerNames: string[];
+  editorNames: string[];
```
Can you rename the file back to `exploration-rights.service.ts` and create a new `exploration-rights-backend-api.service.ts` which only contains the http calls made in this file?

Example: `admin-data.service.ts` and `admin-backend-api.service.ts`.

heyimShivam replied:
Done

heyimShivam committed:
front end complete may e2e test resolved

heyimShivam committed:
test case fro exploration-rights done

heyimShivam committed:
minor change left i.e. variable update (overal all file completed)

heyimShivam committed:
lint check

heyimShivam committed:
Merge remote-tracking branch 'upstream/develop' into exploration-rights

heyimShivam committed:
lint issue

heyimShivam committed:
rename

heyimShivam committed:
restored

heyimShivam committed:
useless nits

heyimShivam committed:
rootScope Apply

heyimShivam committed:
change in frontend test case solved 100 percent cover

heyimShivam commented:
@ashutoshc8101 PTAL

heyimShivam commented:
@srijanreddy98  PTAL

ashutoshc8101 commented on core/templates/pages/exploration-editor-page/services/exploration-rights-backend-api.service.ts:
```diff
@@ -0,0 +1,120 @@
+  }
+
+  async publishPutData(
+      url: string, makePublic: boolean
+  ): Promise<ExplorationRightsBackendData> {
```
Url should be local members in the ExplorationRightsBackendApiService, not passed as function parameters.
And also refrain from writing your or anyone name's in code.

heyimShivam replied:
` Url should be local members in the ExplorationRightsBackendApiService ` can you explain a bit more about this

heyimShivam replied:
Done

heyimShivam committed:
url changes

heyimShivam commented:
@ashutoshc8101  PTAL

heyimShivam committed:
auto indent remove

heyimShivam committed:
Merge remote-tracking branch 'upstream/develop' into exploration-rights

heyimShivam committed:
try to solve lint issue

srijanreddy98 committed:
Fixes part of #9749: Migrate few instances of angular-html-bind (#14263)

* Revert "Revert "Fix part of #9749: Create a component that shows rte-output-display and use it instead of angular-html-bind. (#13229)" (#13361)"

This reverts commit 792de4596996527a696d0041e0eff2b8384eda30.

* Fix issues

* Fix spelling mistakes in comment

* Merge with upstream/develop

* Address review comments:
  1. Move oppia-rte-output.spec from the list of ignored files for innerHTML check in eslintrc and move to inline disables.
  2. Change the CODEOWNER listing from oppia-rte-parser.service.ts to oppia-rte-parser.service*.ts to include the spec file.
  3. Add refernce to the note at the top of file for all value.replace ops.
  4. Miscellaneous changes.

Co-authored-by: Vasa Srijan Reddy <redvasa@amazon.com>

heyimShivam committed:
Merge remote-tracking branch 'upstream/develop' into exploration-rights

heyimShivam committed:
testing improved

heyimShivam committed:
same name

srijanreddy98 committed:
Merge branch 'develop' into exploration-rights

heyimShivam committed:
one more serive mig

heyimShivam committed:
Merge remote-tracking branch 'upstream/develop' into exploration-rights

heyimShivam committed:
Merge remote-tracking branch 'upstream/develop' into exploration-rights

ashutoshc8101 commented on core/templates/pages/exploration-editor-page/services/exploration-rights-backend-api.service.ts:
```diff
@@ -0,0 +1,138 @@
+    private http: HttpClient,
+  ) {}
+
+  async makeCommunityOwnedPutData(
+      explorationId: string, version: number, makeCommunityOwned: boolean):
+    Promise<ExplorationRightsBackendData> {
```
```suggestion
  async makeCommunityOwnedPutData(
      explorationId: string,
      version: number,
      makeCommunityOwned: boolean
    ): Promise<ExplorationRightsBackendData> {
```

heyimShivam replied:
Done

ashutoshc8101 commented on core/templates/pages/exploration-editor-page/services/exploration-rights-backend-api.service.ts:
```diff
@@ -0,0 +1,138 @@
+  }
+
+  async saveRoleChangesPutData(
+      explorationId: string, version: number, newMemberRole: string,
+      newMemberUsername: string): Promise<ExplorationRightsBackendData> {
```
```suggestion
      explorationId: string,
      version: number,
      newMemberRole: string,
      newMemberUsername: string
    ): Promise<ExplorationRightsBackendData> {
```

heyimShivam replied:
Done

ashutoshc8101 commented on core/templates/pages/exploration-editor-page/services/exploration-rights-backend-api.service.ts:
```diff
@@ -0,0 +1,138 @@
+  async saveRoleChangesPutData(
+      explorationId: string, version: number, newMemberRole: string,
+      newMemberUsername: string): Promise<ExplorationRightsBackendData> {
+    const requestUrl = ('/createhandler/rights/' + explorationId);
```
```suggestion
    const requestUrl = '/createhandler/rights/' + explorationId;
```

heyimShivam replied:
Done

ashutoshc8101 commented on core/templates/pages/exploration-editor-page/services/exploration-rights.service.spec.ts:
```diff
@@ -301,8 +317,8 @@ describe('Exploration rights service', function() {
-  it('should check oldrole of user', function() {
-    var sampleDataResultsCopy = angular.copy(sampleDataResults);
+  it('should check oldrole of user', () => {
+    let sampleDataResultsCopy = angular.copy(serviceData);
```
Try not to use anything from the global angular object while migrating a file. The global angular object won't be available once angularjs library is removed from the page. Use `cloneDeep` from lodash instead.

heyimShivam replied:
Done

ashutoshc8101 commented on core/templates/pages/exploration-editor-page/services/exploration-rights.service.ts:
```diff
@@ -17,206 +17,210 @@
+        this.explorationDataService.explorationId,
+        this.explorationDataService.data.version, true)
+      .then((response: ExplorationRightsBackendData) => {
+        let data = response;
```
Is this line necessary, can't we directly use response object here?

heyimShivam replied:
No, Removed ( Done )

ashutoshc8101 reviewed (changes requested):
Just few nits.

heyimShivam committed:
changes made

heyimShivam committed:
Merge remote-tracking branch 'upstream/develop' into exploration-rights

heyimShivam committed:
Merge branch 'exploration-rights' of https://github.com/heyimShivam/oppia into exploration-rights

heyimShivam committed:
done

heyimShivam committed:
some left

heyimShivam commented:
PTAL @ashutoshc8101

heyimShivam committed:
some left

ashutoshc8101 commented on core/templates/pages/exploration-editor-page/services/exploration-rights.service.spec.ts:
```diff
@@ -325,83 +343,99 @@ describe('Exploration rights service', function() {
-      expect(ers.editorNames).toBeUndefined();
-      expect(ers.voiceArtistNames).toBeUndefined();
-      expect(ers.viewerNames).toBeUndefined();
+      expect(ers.ownerNames).toBe(null);
+      expect(ers.editorNames).toBe(null);
```
```suggestion
      expect(ers.ownerNames).toBeNull();
      expect(ers.editorNames).toBeNull();
```

please prefer to use `toBeNull` instead of `toBe(null)` here and elsewhere.

heyimShivam replied:
Done

heyimShivam replied:
`okay` i will use it

ashutoshc8101 reviewed (approved):
LGTM, just one nit.

heyimShivam committed:
to be null

seanlip commented:
Not sure why @krishnarao22 and I are assigned to this, since AFAICT we don't own any files here (even through the teams we're a part of). Taking ourselves off.

EricZLou commented on core/templates/pages/exploration-editor-page/services/exploration-rights-backend-api.service.ts:
```diff
@@ -42,7 +42,9 @@ export class ExplorationRightsBackendApiService {
-      explorationId: string, version: number, makeCommunityOwned: boolean):
+      explorationId: string,
+      version: number,
+      makeCommunityOwned: boolean):
```
If all of the arguments can fit on a single line, let's keep them on a single line! If the linter says it's too long, then we can put arguments on their own lines like you've done here.

heyimShivam replied:
when i put them like this  :
`async makeCommunityOwnedPutData(explorationId: string,  version: number, makeCommunityOwned: boolean)` 
linter is gaving error too long. 

when I put them like this :
`async makeCommunityOwnedPutData(        
                                                                                                                           
    explorationId: string, version: number,  makeCommunityOwned: boolean)`
linter is accepting them

`last method is used above`

`which one should I prefer ?`

EricZLou replied:
Let's keep it the way it was before! 
```
async makeCommunityOwnedPutData(
    explorationId: string, version: number,  makeCommunityOwned: boolean)`
```

heyimShivam replied:
okay

heyimShivam replied:
Done

EricZLou commented on core/templates/pages/exploration-editor-page/services/exploration-rights-backend-api.service.ts:
```diff
@@ -65,9 +70,11 @@ export class ExplorationRightsBackendApiService {
-      explorationId: string, version: number, viewableIfPrivate: boolean
+      explorationId: string,
+      version: number,
+      viewableIfPrivate: boolean
```
Same here.

heyimShivam replied:
same as the above case

EricZLou replied:
Same response!

heyimShivam replied:
okay

heyimShivam replied:
Done

EricZLou reviewed (changes requested):
Small things -- thanks!

heyimShivam commented:
@EricZLou  PTAL

EricZLou commented:
Thanks @heyimShivam I just responded!

heyimShivam committed:
suggest changes

heyimShivam committed:
Merge remote-tracking branch 'upstream/develop' into exploration-rights

heyimShivam commented:
@EricZLou  PTAL

EricZLou reviewed (approved):
LGTM -- thanks @heyimShivam!

EricZLou commented:
Also, does this fix part of #8016?

heyimShivam commented:
this issue deals with both #9749 and #8016. but as I have moved exploration-rights-backend-api. service, ts in this, so I decided to label this issue as #8016

gp201 reviewed (approved):
LGTM for codeowner files

aishwary023 commented on core/templates/pages/exploration-editor-page/services/exploration-rights-backend-api.service.spec.ts:
```diff
@@ -0,0 +1,222 @@
+    httpTestingController.verify();
+  });
+
+  it('should put requestData when makeCommunityOwnedPutData called', fakeAsync(
```
the test description can be improved. You may go through the FE Testing Wiki & read other similar test descriptions

heyimShivam replied:
tried to make the test name more clear ( Done )

aishwary023 commented on core/templates/pages/exploration-editor-page/services/exploration-rights-backend-api.service.spec.ts:
```diff
@@ -0,0 +1,222 @@
+    })
+  );
+
+  it('should put requestData when publishPutData called', fakeAsync(() => {
```
ditto comment for all tests

heyimShivam replied:
tried to make the test name more clear ( Done )

aishwary023 commented on core/templates/pages/exploration-editor-page/settings-tab/settings-tab.component.spec.ts:
```diff
@@ -736,6 +737,32 @@ describe('Settings Tab Component', () => {
       expect(ctrl.isExplorationLockedForEditing()).toBe(false);
     });
 
+    it('should call closeRolesForm in reassignRole', fakeAsync(() => {
```
ditto

heyimShivam replied:
tried to make the test name more clear ( Done )

heyimShivam committed:
changing name

heyimShivam committed:
e2e resolve

heyimShivam committed:
Merge remote-tracking branch 'upstream/develop' into exploration-rights

heyimShivam committed:
naming improved

heyimShivam commented:
@aishwary023 PTAL
